firedrake/ensemble.py

```diff
@@ -35,32 +35,123 @@ def __init__(self, comm, M):
         """
         Allreduce a function f into f_reduced over :attr:`ensemble_comm`.
 
         :arg f: The a :class:`.Function` to allreduce.
         :arg f_reduced: the result of the reduction.
         :arg op: MPI reduction operator.
-        :raises ValueError: if communicators mismatch, or function sizes mismatch.
+        :raises ValueError: if function communicators mismatch each other or the ensemble spatial communicator, or if the functions are in different spaces
         """
-        if MPI.Comm.Compare(f_reduced.comm, f.comm) not in {MPI.CONGRUENT, MPI.IDENT}:
-            raise ValueError("Mismatching communicators for functions")
-        if MPI.Comm.Compare(f.comm, self.comm) not in {MPI.CONGRUENT, MPI.IDENT}:
-            raise ValueError("Function communicator does not match space communicator")
+        self._check_function(f, f_reduced)
+
         with f_reduced.dat.vec_wo as vout, f.dat.vec_ro as vin:
-            if vout.getSizes() != vin.getSizes():
-                raise ValueError("Mismatching sizes")
             self.ensemble_comm.Allreduce(vin.array_r, vout.array, op=op)
         return f_reduced
 
-    def __del__(self):
-        if hasattr(self, "comm"):
-            self.comm.Free()
-            del self.comm
-        if hasattr(self, "ensemble_comm"):
-            self.ensemble_comm.Free()
-            del self.ensemble_comm
+    def iallreduce(self, f, f_reduced, op=MPI.SUM):
+        """
+        Allreduce (non-blocking) a function f into f_reduced over :attr:`ensemble_comm`.
+
+        :arg f: The a :class:`.Function` to allreduce.
+        :arg f_reduced: the result of the reduction.
+        :arg op: MPI reduction operator.
+        :returns: list of MPI.Request objects (one for each of f.split()).
+        :raises ValueError: if function communicators mismatch each other or the ensemble spatial communicator, or if the functions are in different spaces
+        """
+        self._check_function(f, f_reduced)
+
+        return [self.ensemble_comm.Iallreduce(fdat.data, rdat.data, op=op)
+                for fdat, rdat in zip(f.dat, f_reduced.dat)]
+
+    def reduce(self, f, f_reduced, op=MPI.SUM, root=0):
+        """
+        Reduce a function f into f_reduced over :attr:`ensemble_comm` to rank root
+
+        :arg f: The a :class:`.Function` to reduce.
+        :arg f_reduced: the result of the reduction on rank root.
+        :arg op: MPI reduction operator.
+        :arg root: rank to reduce to
+        :raises ValueError: if function communicators mismatch each other or the ensemble spatial communicator, or is the functions are in different spaces
+        """
+        self._check_function(f, f_reduced)
+
+        # need to use `vec` not `vec_wo` for f_reduced otherwise function will be blanked out
+        # when rank!=root because `vec_wo` doesn't copy over existing data into the pyop2 vector
+        with f_reduced.dat.vec as vout, f.dat.vec_ro as vin:
```

Comment:
You should be able to pass `None` on all the other ranks (since `vout` isn't needed on those). But that suggests adding some new `.local_vec` properties in PyOP2. So meh.

Reply:
What would the `.local_vec` properties be / why would they be needed? I haven't used that before

Reply:
I can avoid using `.vec` if that's preferable, it does avoid the extra copy out on root
```
if self.ensemble_comm.rank == root:
    with f_reduced.dat.vec_wo as vout, f.dat.vec_ro as vin:
        self.ensemble_comm.Reduce(vin.array_r, vout.array, op=op, root=root)
else:
    with f.dat.vec_ro as vin:
        self.ensemble_comm.Reduce(vin.array_r, None, op=op, root=root)
```

Reply:
I think that doesn't work because `.vec_wo` is collective. `local_vec` would do the work of gluing together the pieces of a function, but give you a vec on `COMM_SELF` rather than `comm` (so it wouldn't be collective).

Thinking about it though, that probably is bad because you would get out of whack with state tracking.

Reply:
Collective over which communicator?
This version passes the tests (although that doesn't mean it isn't undefined behaviour)

Reply:
`f.dat.vec_XX` is collective over the communicator of `f.dat`

Reply:
I think this is ok then, because `f_reduced` is defined over `ensemble.comm`, so `f_reduced` on any `rank!=root` does not "know" that the `f_reduced` on `rank==root` is constructing a `vec_wo`,

Reply:
One thing to check is to ask (after doing this) on all ranks that share `f_reduced` that `.dat.vec.stateGet()` is the same.

Reply:
It would appear that they are. Would you be happy with this version if I add a test for the stateGet?

Reply:
Yes, thanks.

Reply:
I've added this test now. I've not used `stateGet` before so please let me know if I'm accessing it incorrectly.
I've also updated the `Ensemble` section of the parallelism docs to reflect what methods we now implement.